### PR TITLE
Refactors SM Dusting into a Component

### PR DESCRIPTION
**Pull request by DragonTrance**
## About The Pull Request
Refactors the SM dusting into a component, and now can be added to anything! The SM still does interactions itself because of the special text it gives, but it lets the component handle the dusting. This isn't the case for the SM sword and instead is (near) completely relied on the component now.

## Why It's Good For The Game
Mainly for admemery. Also very slightly changes how logging texts are displayed to the admins when something gets consumed for propurr gremmer

Also this:

https://user-images.githubusercontent.com/26515331/149614661-7e480fb0-4a34-4119-a676-6e454c37a918.mp4

## Changelog

:cl:
refactor: Supermatter dusting is now a component and can be added to any atom. **ANY** atom.
add: The SM sword now has a chance of dusting you when picking it up after it's been dropped.
code: Removed incredibly useless appearances of \the in SM code.
code: Replaced src.loc with just loc in SM sword code
admin: SM dusting gives slightly better grammar to admins and investigation logs
/:cl:

### CODE REVIEW

File: code/datums/components/dusting.dm [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	/// Audible text to show to people when they can't see what's being consumed, but are still in range to recieve consumtion messages. %parent% for parent this component is attached to, %target% to what got consumed
+	var/consume_text_unseen = span_hear("An unearthly ringing fills your ears, and you find your skin covered in new radiation burns.")
+	/// If a target has this flag, don't consume it
+	var/ignore_flags = SUPERMATTER_IGNORES_1
+	/// If a mob has this flag, don't consume it, along with ignore_flags
+	var/ignore_status_flags = GODMODE
+	/// Ignore subtypes of this type to consume. Becomes typecache list
+	var/list/ignore_subtypesof = list(/obj/effect, /obj/item/scalpel/supermatter)
+
+	/// Currently not sending any messages to admins to prevent chat spam. Not a bool
```
Comment (LemonInTheDark) [inline code/datums/components/dusting.dm:40]: Well what is it then?

File: code/datums/components/dusting.dm [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	var/ignoring_admin_messages = 0
+	/// Skipping retype vars
+	var/atom/atom_parent
```
Comment (LemonInTheDark) [inline code/datums/components/dusting.dm:43]: This is a bad pattern, don't do it

File: code/modules/power/supermatter/supermatter.dm [hunk 3]
```diff
@@ -1111,44 +1103,23 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
...
-		if(ishuman(near_mob) && SSradiation.wearing_rad_protected_clothing(near_mob))
-			continue
-		if(near_mob in view())
-			near_mob.show_message(span_danger("As \the [src] slowly stops resonating, you find your skin covered in new radiation burns."), MSG_VISUAL,
-				span_danger("The unearthly ringing subsides and you find your skin covered in new radiation burns."), MSG_AUDIBLE)
-		else
-			near_mob.show_message(span_hear("An unearthly ringing fills your ears, and you find your skin covered in new radiation burns."), MSG_AUDIBLE)
+	var/datum/component/dusting/dust = GetComponent(/datum/component/dusting)
+	dust?.consume_atom(consumed_object)
+
```
Comment (LemonInTheDark) [inline code/modules/power/supermatter/supermatter.dm:1122]: The fuck is this, why is this needed?

File: code/modules/power/supermatter/supermatter.dm [hunk 4]
```diff
@@ -1091,17 +1093,7 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
...
-			span_hear("You hear an unearthly noise as a wave of heat washes over you."))
-	else if(isobj(hit_object) && !iseffect(hit_object))
-		hit_object.visible_message(span_danger("\The [hit_object] smacks into \the [src] and rapidly flashes to ash."), null,
-			span_hear("You hear a loud crack as you are washed with a wave of heat."))
-	else
-		return
-
-	playsound(get_turf(src), 'sound/effects/supermatter.ogg', 50, TRUE)
+	. = ..()
 	Consume(hit_object)
```
Comment (LemonInTheDark) [inline code/modules/power/supermatter/supermatter.dm:1097]: What does the consume var do now?

File: code/game/objects/items/melee/misc.dm [hunk 5]
```diff
@@ -184,91 +184,41 @@
...
 	if(!isturf(src.loc))
 		var/atom/target = src.loc
 		forceMove(target.loc)
-		consume_everything(target)
+		dust?.consume_atom(target)
 	else
 		var/turf/turf = get_turf(src)
 		if(!isspaceturf(turf))
-			consume_turf(turf)
+			dust?.consume_atom(turf)
```
Comment (LemonInTheDark) [inline code/game/objects/items/melee/misc.dm:207]: Also no

File: code/datums/components/dusting.dm [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+			near_mob.show_message(visual_message, MSG_VISUAL, audible_message, MSG_AUDIBLE)
+		else
+			near_mob.show_message(unseen_message, MSG_AUDIBLE)
+		atom_parent.investigate_log("has irradiated [key_name(near_mob)] after consuming [consumed_atom].", INVESTIGATE_SUPERMATTER)
+
+//Helper proc
+/datum/component/dusting/proc/rename_text(text, target)
+	. = replacetext(text, "%parent%", "[atom_parent]")
+	if(target)
+		return replacetext(., "%target%", "\The [target]")
```
Comment (LemonInTheDark) [inline code/datums/components/dusting.dm:249]: What is this for?

File: code/datums/components/dusting.dm [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+	//Next, consume the actual atom
+	if(ismob(consumed_atom))
+		var/mob/living/consumed_living = consumed_atom
+		if(consumed_living.status_flags & ignore_status_flags)
+			return
+		message_admins("[atom_parent] ([src.type]) has consumed [key_name_admin(consumed_living)] [ADMIN_JMP(atom_parent)].")
+		atom_parent.investigate_log("has consumed [key_name(consumed_living)].", INVESTIGATE_SUPERMATTER)
+		consumed_living.visible_message(span_danger("[consumed_living] slams into [atom_parent] inducing a resonance... [consumed_living.p_their()] body starts to glow and burst into flames before flashing into dust!"),
+			span_userdanger("You slam into [atom_parent] as your ears are filled with unearthly ringing. Your last thought is \"Oh, fuck.\""),
+			span_hear("You hear an unearthly noise as a wave of heat washes over you."))
```
Comment (LemonInTheDark) [inline code/datums/components/dusting.dm:184]: This isn't very modular

File: code/datums/components/dusting.dm [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+	for(var/mob/living/near_mob in range(consume_text_range, parent))
+		if(HAS_TRAIT(near_mob, TRAIT_RADIMMUNE) || issilicon(near_mob))
+			continue
+		if(ishuman(near_mob) && SSradiation.wearing_rad_protected_clothing(near_mob))
+			continue
+		if(near_mob in view(consume_text_range, atom_parent))
+			near_mob.show_message(visual_message, MSG_VISUAL, audible_message, MSG_AUDIBLE)
+		else
+			near_mob.show_message(unseen_message, MSG_AUDIBLE)
+		atom_parent.investigate_log("has irradiated [key_name(near_mob)] after consuming [consumed_atom].", INVESTIGATE_SUPERMATTER)
```
Comment (LemonInTheDark) [inline code/datums/components/dusting.dm:243]: Why are we doing all this here, and not by sending a signal -> Hooking into that signal on the objects that want to do radiation and well, doing radiation

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
You're using GetComponent a lot, which is never a great sign.
A lot of this seems like very supermatter specific behavior, I'm not sure it's a great thing to generalize. Is this just for admin memes?

File: code/modules/power/supermatter/supermatter.dm (hunk 3 above)
Comment (DragonTrance) [reply code/modules/power/supermatter/supermatter.dm:1122]: How do I avoid this because people keep thinking nasty things about the getcomponent and then calling a proc on it

File: code/datums/components/dusting.dm (hunk 1 above)
Comment (DragonTrance) [reply code/datums/components/dusting.dm:40]: Its a var that ranges from 0 to 2 cause I'm too lazy for defines. It's mostly for logging when the component is allowed to consume all contents of a target and doesn't flood chat

File: code/modules/power/supermatter/supermatter.dm (hunk 4 above)
Comment (DragonTrance) [reply code/modules/power/supermatter/supermatter.dm:1097]: Calls the component consuming and adds power to the SM in a regular way

File: code/datums/components/dusting.dm (hunk 7 above)
Comment (DragonTrance) [reply code/datums/components/dusting.dm:184]: It will soontm become modular because you mentioned that

File: code/datums/components/dusting.dm (hunk 6 above)
Comment (DragonTrance) [reply code/datums/components/dusting.dm:249]: > You're using GetComponent a lot, which is never a great sign. A lot of this seems like very supermatter specific behavior, I'm not sure it's a great thing to generalize. Is this just for admin memes?

Mostly for admemes, yeah. I dont really see where future code will use this but hey, we have organs that make you puke slimes, at some point this component will be used outside admemery and the sm sword

**Comment by DragonTrance** [comment]
> You're using GetComponent a lot, which is never a great sign. A lot of this seems like very supermatter specific behavior, I'm not sure it's a great thing to generalize. Is this just for admin memes?

Mostly for admemery yeah. No, this isn't a great thing to generalize, but you never know what xould be added in the future. We got organs that make you puke slimes, at some point this component will be used outside sm-related things.

More is always better in its own way, unless the more makes the server crash uncontrollably

**Comment by DragonTrance** [comment]
Making this more modular and more "generalized" by adding callback options and other stuff like the food component

File: code/datums/components/dusting.dm (hunk 2 above)
Comment (tralezab) [reply code/datums/components/dusting.dm:43]: agree with potato on this

File: code/datums/components/dusting.dm (hunk 1 above)
Comment (tralezab) [reply code/datums/components/dusting.dm:40]: then write that in the documentation, ya?

File: code/datums/components/dusting.dm [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+		RegisterSignal(parent, COMSIG_ITEM_PICKUP, .proc/on_pickup)
+		RegisterSignal(parent, COMSIG_MOVABLE_IMPACT, .proc/on_throw_impact)
+		RegisterSignal(parent, COMSIG_ATOM_BUMPED, .proc/on_bumped)
+		RegisterSignal(parent, list(\
+			COMSIG_ATOM_ATTACK_HAND,
+			COMSIG_ATOM_HULK_ATTACK,
+			COMSIG_ATOM_ATTACK_ANIMAL,
+			COMSIG_ATOM_ATTACK_BASIC_MOB,
+			COMSIG_ATOM_ATTACK_PAW
+			), .proc/on_attack_hand)
```
Comment (tralezab) [inline code/datums/components/dusting.dm:101]: unindent this line

File: code/datums/components/dusting.dm [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+		COMSIG_PARENT_ATTACKBY,
+		COMSIG_ITEM_PICKUP,
+		COMSIG_MOVABLE_IMPACT,
+		COMSIG_ATOM_BUMPED,
+		COMSIG_ATOM_ATTACK_HAND,
+		COMSIG_ATOM_HULK_ATTACK,
+		COMSIG_ATOM_ATTACK_ANIMAL,
+		COMSIG_ATOM_ATTACK_BASIC_MOB,
+		COMSIG_ATOM_ATTACK_PAW,
+		))
```
Comment (tralezab) [inline code/datums/components/dusting.dm:117]: unindent this line

File: code/modules/power/supermatter/supermatter.dm (hunk 3 above)
Comment (tralezab) [reply code/modules/power/supermatter/supermatter.dm:1122]: have a signal for this, yeah?

**Review by tralezab — CHANGES_REQUESTED** [review]
mostly just documentation stuff

**Comment by DragonTrance** [comment]
> have a signal for this, yeah?
For some reason can't reply to that specific message, but
I'm handling this with utilizing callbacks, because I find it excessive trying to send signals from a component like this. Signals would be fine if maybe there were two components that did something very similar

**Comment by LemonInTheDark** [comment]
?
